Dwarf Fortress updates

abbradar opened the pull request:
This is a massive rework of the current dwarf_fortress package.
The main reason for rework is to split game, its libraries and wrappers,
enabling builds on Hydra and general maintainability.

dfhack, dwarf-fortress-unfuck, dwarf-fortress-original and dwarf-fortress
are now all separate derivations. All were updated to the latest versions
supporting DF 0.42.04 (some things taken from development branches).

dfhack got a patch for using more system libraries.

Wrapper scripts were also redone -- now they support DF_DIR variable (for
setting another data directory) and try to save more space by means of
symlinking. They now try to be more smart, updating those directories that were
not touched by user automatically.

All the new packages now reside in dwarf-fortress-packages. dwarf_fortress was
renamed into dwarf-fortress per our conventions, old name is deprecated. Also
duplicate dfhack derivation is removed.

a1russell committed:
dwarf-fortress: add phoebus

abbradar committed:
dwarf-fortress: 0.40.24 -> 0.42.05, split derivations

This is a massive rework of the current dwarf_fortress package.
The main reason for rework is to split game, its libraries and wrappers,
enabling builds on Hydra and general maintainability.

dfhack, dwarf-fortress-unfuck, dwarf-fortress-original and dwarf-fortress
are now all separate derivations. All were updated to the latest versions
supporting DF 0.42.04 (some things taken from development branches):

dfhack: 20150927 -> 20160118
dwarf-fortress-unfuck: 0.40.24 -> 20160118
dwarf-fortress-original: 0.40.24 -> 0.42.05

dfhack got a patch for using more system libraries.

Wrapper scripts were also redone -- now they support DF_DIR variable (for
setting another data directory) and try to save more space by means of
symlinking. They now try to be more smart, updating those directories that were
not touched by user automatically.

All the new packages now reside in dwarf-fortress-packages. dwarf_fortress was
renamed into dwarf-fortress per our conventions, old name is deprecated. Also
duplicate dfhack derivation is removed.

Also drop dwarfFortress.enableDFHack config option; it can be re-added later
but now is deemed controversial.

abbradar committed:
dwarf-therapist: 31.0.0 -> 36.0.0

abbradar commented:
(For some reason I keep mixing up "has" and "needs" tags, sorry)

a1russell commented:
It's not obvious to me the difference between, e.g. `dwarf-fortress` and `dwarf-fortress-original`. Could you please explain to me?

abbradar commented:
`dwarf-fortress-original` is the game as distributed by Tarn Adams (with some `patchelf` on top). `dwarf-fortress` is a wrapper, which takes the game, `dfhack` (if needed) and packs a bunch of shell scripts that create `~/.local/share/df_linux`, set up everything and run DF (from the first derivation).

a1russell commented:
Had some time today to try this out. Definitely seems like a step in the right direction. I don't have the Nix language prowess to do this myself, yet, for sure.

Regarding the new symlinking approach that these changes take, I'd like to discuss with you what I see as two alternatives.

First, [you mentioned that it might be possible to get rid of the symlinks and/or copied filesin the home directory altogether](https://github.com/NixOS/nixpkgs/pull/11301#issuecomment-171020200). Do you think it would be worth it to do that? Personally, after realizing the `save` directory is located there, and that installing themes and dfhack is much easier with the symlinked/copied files, I don't think so, but I'd like to hear your opinion.

Second, the difference between the current approach, before this pull request, and the approach this pull request takes, is that this pull request has more symbolic links in the data directory rather than copied directories. I am having trouble thinking how to install a theme with this new approach.

I mean, it seems to me like, when creating a package for a theme, we could link to the raw, data/art, data/config and such directories of the theme, but how do we get the dwarf-fortress script to link to the theme's versions of these directories rather than the vanilla versions?

The approach I took in #11301 was to install the theme files over the top of the original files in the install phase, but I don't think that meshes with this new approach, either.

a1russell commented:
Also, you can add me as a maintainer for dfhack (see #10800).

abbradar committed:
phoebus-theme: update for 0.42.05, make a derivation

Also specify license, see https://github.com/fricy/Phoebus/issues/5

abbradar committed:
dwarf-fortress-wrapper: add themes support

Theme can be specified either as a derivation or as a string, in which
case it will be taken by name from a pre-defined set of themes available in
nixpkgs.

abbradar commented:
I've added you as a maintainer for DFHack, cherry-picked your Phoebus theme and converted it to a derivation and added `extraPackages` argument for the DF wrapper. As you can see, it results in pretty much no extra code (rather, it unifies `dfhack` handling with that of themes and `dwarf-fortress` itself) and uses our `buildEnv` function to do all the heavy lifting.

I think that we may just leave it as-is w.r.t. symlinks -- after some thought, getting rid of them would be a tough job, and would likely require patch for `libgraphics` (assuming it is the library DF uses for IO; if it does it by itself, there's no way besides asking Toady to add necessary support).

Regarding the second question: any new theme (or any other "replace DF files" thing) can be just packaged as an ordinary derivation (using `stdenv.mkDerivation`), also exposing DF version it supports via `passthru.dfVersion`. The result should be bunch of files and directories in `$out` which are to be overlayed on (i.e. `data`, `raw` etc.).  Then you can add it to `extraPackages` of the wrapper and that's it!

An example for Dwarf Fortress with DFHack and Phoebus theme (test it with `nix-build test.nix`):
```
{ pkgs ? import ./. { } }:
pkgs.dwarf-fortress.override {
  enableDFHack = true;
  extraPackages = [ pkgs.dwarf-fortress-packages.phoebus-theme ];
}
```

a1russell commented:
Thanks @abbradar! I'll try this out ASAP. I don't completely understand all the moving parts, yet, but I'll get there. I think I need to especially read your commit on implementing `extraPackages`.

Only thing I think might be nice now is adding a config for phoebus similar to `config.dwarfFortress.enableDFHack`, but if that's tricky, I don't think it's too important.

abbradar commented:
IMO we don't need it -- adding separate configuration argument for each theme will hurt and why do so when we have a generic argument "give me a theme to use"? DFHack is special here because it adds a separate launch script `dfhack` to the wrapper.

Either way we should document somewhere that `extraPackages` can be used for adding a theme on top. I don't quite know where we should document such things; any ideas?

Feel free to ping me about any parts that you don't get -- I'll be happy to explain (and make them more understandable, if needed).

a1russell commented:
I think the documentation part is my concern, as well, @abbradar. Personally I find it fairly easy to read `all-packages.nix` to find out what kind of configuration options are available, but there doesn't seem to be a good way to discover features only available via overrides.

So while I do sort of disagree with you that the config item for phoebus isn't needed, I'll admit to being selfish enough to be happy that this will work for me since **I** know about it.

Will make sure to ping you with questions as I read through. Thanks!

abbradar commented:
Hm, what do you think about making it instead a `nullOr package` option?
e.g. `extraPackages = [ dwarf-fortress-packages.phoebus-theme ]` -> `theme = dwarf-fortress-packages.phoebus-theme`?

abbradar commented:
I've fixed few issues regarding `data/announcement` needed to be writable and cleaned some parts. Let me know what do you think about `theme` argument instead of `extraPackages` -- I think it is pretty self-documenting this way. We lose some flexibility (`extraPackages` allowed to specify several packages which would be overlapped) but I doubt it would be needed by somebody at all.

a1russell commented:
I hadn't heard of `nullOr` before, so are you referring to something like the [`package` option of the awesome package](https://github.com/NixOS/nixpkgs/blob/release-15.09/nixos/modules/services/x11/window-managers/awesome.nix#L29-L34)? If so, yeah, I like that a *lot* better.

abbradar commented:
Check this out, you can now use it like this:
```
dwarf-fortress.override {
  enableDFHack = true;
  theme = dwarf-fortress-packages.phoebus-theme;
}
```

a1russell commented:
@abbradar I like this a bit better than the `extraPackages`, although you had my hopes up after looking at that `nullOr` option, haha. :wink: 

I just tried this, and it works great. I'm still working to understand how the build works, exactly, though.

One thing I've noticed, is we are dumping files right to the root of the nix store, and using them. Is that acceptable? It doesn't seem conventional. For example, `/nix/store/72hrmjjyfjbzd6vvj0x0y7rb57q7bxwc-dwarf-fortress-0.42.05/bin/dwarf-fortress` sources `/nix/store/81gradjp85k0by33nid57jzyf73dlwsy-dwarf-fortress-init.in`.

By the way, those *.in files I named as such to signify that they were templates to be processed via the `substitute` command. Probably any file that's no longer a template or already processed should be renamed.

abbradar commented:
@a1russell Oh, we misunderstood each other -- I just wanted to convey the type of the resulting option. Sorry!

I don't know any better way to do this though -- we may indeed add `enablePhoebus` flag but
```
dwarf-fortress.override {
  enablePhoebus = true;
}
```
doesn't seem much better to me than
```
dwarf-fortress.override {
  theme = dwarf-fortress-packages.phoebus-theme;
}
```
and I expect we would have more themes in future (Ironhand, for example).

Regarding usage of files -- this is perfectly acceptable behavior, we do this in several places, mostly in NixOS for various configuration files.

I'll fix `dwarf-fortress-init.in` name, thanks for spotting!

a1russell commented:
@abbradar What are `passthru` and `preferLocalBuild`? I can't find anything that explains them well enough for me.

abbradar commented:
`passthru` is an attribution set which doesn't affect a build in any way (i.e., it isn't visible inside the build), but is accessible later as if it was set inside the derivation itself.

For example,
```
stdenv.mkDerivation {
  name = "hello";
  src = ....
  buildInputs = [ a b c ];
  buildCommand = ''
    mv $src $out
  '';
  passthru = {
    value1 = 4;
    value2 = [ "foo" "bar" ];
  };
}
```

Here, if I change `buildInputs` or `src`, the derivation would be rebuilt (because something has changed). But I can change anything in `passthru` to the heart's content (add a value, modify or remove) and the rebuild will not happen, because values from `passthru` are not visible the build. However, if I call this derivation:

```
  hello = callPackage ./hello.nix { }
```

I can get the values inside with `hello.value1`, `hello.value2` etc (notice absence of `passthru`).

You may notice that it's very similar to `meta` -- that's right. However, there are two differences. First, technical -- you access `meta` values from outside the derivation with `hello.meta.platforms`, not `hello.platforms`. Second, semantical -- `passthru` is used for arbitrary values maintainer thinks would be useful for people, while `meta` has predefined fields which are expected to be filled.

I use `passthru.dfVersion` to pass through DF version in convenient form in `dwarf-fortress-original` and to set DF version for which a package works (e.g. current `dfhack` works with `0.42.05` but it might have been only for `0.42.04`). Later, in the wrapper, I check if each package that is used (theme and `dfhack` at the moment) has the same exported version as current original Dwarf Fortress.

`preferLocalBuild` is a flag that tells Nix that the package is so lightweight in terms of build (i.e. copy several files from $src to $out or write a text file, like `substituteAll`) that there's no need to look for it in binary caches -- it's faster to just build it locally. It also tells Hydra that this package doesn't need to be exported as binary cache (noone would use it, after all).

abbradar commented:
I've added some documentation at https://github.com/NixOS/nixpkgs/commit/f727f815cb075f6ff466f8c8a02a275502709eec and https://github.com/NixOS/nixpkgs/pull/12524

a1russell commented:
Perfect, thanks for the great explanation. I'm a lot closer to understanding everything now. I hope to have some time to sit down and go through the build steps manually this weekend. Thanks for your patience!

abbradar commented:
No problem, feel free to ping me further!

a1russell commented:
One thing that can be cleaned up is that there are currently two md5 files. One of them is a trimmed one that dwarf-therapist uses. But later, when I added dfhack, I added the second, full one since dfhack needs it. I just hadn't got around to refactoring dwarf-therapist to use the full one instead, since it can just trim it.

a1russell commented:
Other than that (and the merge conflict?), I've reviewed the changes, and I think I understand most of what's going on now! I think this is good. I'd be glad to see this merged.

a1russell commented:
One last thing to follow up on, since we were discussing it earlier in the thread: the possibility of a config item for theme. I did like how I was able to have:
```
dwarfFortress = {
  enableDFHack = true;
  theme = "phoebus";
};
```
It would still be possible with a set that maps the string to a package, e.g. `themePkgsByName = { "phoebus" = phoebus-theme }`, right? Then we could set the actual theme to whatever the value is, given the key that is the config item.

But if this is too complex for your taste, I understand.

abbradar committed:
dwarf-fortress: remove short md5 sums

abbradar commented:
Actually, while I see the string-specifier idea as somewhat redundant, it seems very much fine to me! I've implemented it so that you can use either a package or a string as `theme` argument -- check it out. I've also redone checksums thing and rebased atop current `master`.

a1russell commented on pkgs/top-level/all-packages.nix:
```diff
@@ -14104,23 +14104,17 @@ let
+  dwarf-fortress-packages = callPackage ../games/dwarf-fortress { };
+
+  dwarf-fortress = dwarf-fortress-packages.dwarf-fortress.override {
     enableDFHack = config.dwarfFortress.enableDFHack or false;
```
Also add `theme = config.dwarfFortress.theme or null;`

abbradar replied:
Oh, I completely forgot about that part.

AFAIK we have a policy of avoiding `config` options in favor of package
overrides, especially when an option doesn't affect multiple packages
(like `pulseaudio`). However, we already have one for DFHack. So we have
three choices:

1. Remove existing option, breaking some setups but consistently getting
rid of it;
2. Leave it as-is as a message that we don't support config options more
and hope to deal with this later;
3. Add a new option, getting consistency and compatibility but breaking
the policy even more.

I think now is a good moment to make choices so let's avoid (2). This
leaves us with (1) vs (3). Personally I'm a little in favor of dropping
options -- we are renaming the package as well, breaking compatibility
already. What's your opinion on this?

On January 24, 2016 8:14:47 AM GMT+03:00, Adam Russell
<notifications@github.com> wrote:

    In pkgs/top-level/all-packages.nix
    <https://github.com/NixOS/nixpkgs/pull/12339#discussion_r50630354>:

    >    dhewm3 = callPackage ../games/dhewm3 {};
    >  
    >    drumkv1 = callPackage ../applications/audio/drumkv1 { };
    >  
    > -  dwarf_fortress = callPackage_i686 ../games/dwarf-fortress {
    > -    SDL_image = pkgsi686Linux.SDL_image.override {
    > -      libpng = pkgsi686Linux.libpng12;
    > -    };
    > -    inherit (pkgsi686Linux.perlPackages) XMLLibXML XMLLibXSLT;
    > +  dwarf-fortress-packages = callPackage ../games/dwarf-fortress { };
    > +
    > +  dwarf-fortress = dwarf-fortress-packages.dwarf-fortress.override {
    >      enableDFHack = config.dwarfFortress.enableDFHack or false;

    Also add |theme = config.dwarfFortress.theme or null;|

    —
    Reply to this email directly or view it on GitHub
    <https://github.com/NixOS/nixpkgs/pull/12339/files#r50630354>.

a1russell replied:
Ah, I wasn't aware of that policy. I think I was emulating the `chromium` and `wine` package when I added those config options. Might I ask the rationale behind the policy? Is it documented somewhere? I rather liked the config since it makes it plain how the package can be configured. It's nice to just look in `all-packages.nix` and see the possible config items right there.

That said, I'd rather not break a policy just because I don't agree with it. Let's go with option 1.

abbradar replied:
Hm, now that you pointed that out, I can't find any definitive source. I've filed https://github.com/NixOS/nixpkgs/issues/12596, let's wait a bit.

a1russell replied:
Thanks for raising that @abbradar. I'll watch to see how the discussion turns out. And thanks for already communicating my opinion.

abbradar replied:
It feels this discussion would be long and the policy actually isn't as agreed upon as I've thought. I propose to resolve this as if the policy doesn't exist using your judgement as the current /de facto/ package maintainer. You would prefer the additional `config` variable to be added in this case, correct?

a1russell replied:
Since there's no agreement, I believe we should still go with option 3, since we can add the config options in a separate PR without breaking compatibility.

a1russell commented:
Once that line is added, I think this is done!

abbradar commented:
Rebased, squashed several things into less commits and merged. Thanks!
